ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/table/LoadPartitions.java

```diff
@@ -0,0 +1,259 @@
+        + partSpecToString(partSpec.getPartSpec()) + " with source location: "
+        + partSpec.getLocation());
+    Path tmpPath = context.utils.getExternalTmpPath(replicaWarehousePartitionLocation);
+
+    Task<?> copyTask = ReplCopyTask.getLoadCopyTask(
+        event.replicationSpec(),
+        sourceWarehousePartitionLocation,
+        tmpPath,
+        context.hiveConf
+    );
+
+    Task<?> addPartTask = TaskFactory.get(
+        new DDLWork(new HashSet<>(), new HashSet<>(), addPartitionDesc),
+        context.hiveConf
+    );
+
+    Task<?> movePartitionTask = movePartitionTask(table, partSpec, tmpPath);
+
+    copyTask.addDependentTask(addPartTask);
+    addPartTask.addDependentTask(movePartitionTask);
+    return copyTask;
+  }
+
+  /**
+   * This will create the move of partition data from temp path to actual path
+   */
+  private Task<?> movePartitionTask(Table table, AddPartitionDesc.OnePartitionDesc partSpec,
+      Path tmpPath) {
+    LoadTableDesc loadTableWork = new LoadTableDesc(
+        tmpPath, Utilities.getTableDesc(table), partSpec.getPartSpec(),
+        event.replicationSpec().isReplace()
+    );
+    loadTableWork.setInheritTableSpecs(false);
+    MoveWork work = new MoveWork(new HashSet<>(), new HashSet<>(), loadTableWork, null, false);
+    return TaskFactory.get(work, context.hiveConf);
+  }
+
+  private Path locationOnReplicaWarehouse(Table table, AddPartitionDesc.OnePartitionDesc partSpec)
+      throws MetaException, HiveException, IOException {
+    String child = Warehouse.makePartPath(partSpec.getPartSpec());
+    if (tableDesc.getLocation() == null) {
+      if (table.getDataLocation() == null) {
+        Database parentDb = context.hiveDb.getDatabase(tableDesc.getDatabaseName());
+        return new Path(
+            context.warehouse.getDefaultTablePath(parentDb, tableDesc.getTableName()), child);
+      } else {
+        return new Path(table.getDataLocation().toString(), child);
+      }
+    } else {
+      return new Path(tableDesc.getLocation(), child);
+    }
+  }
+
+  private Task<? extends Serializable> alterSinglePartition(AddPartitionDesc desc,
+      ReplicationSpec replicationSpec, Partition ptn) {
+    desc.setReplaceMode(true);
+    if ((replicationSpec != null) && (replicationSpec.isInReplicationScope())) {
+      desc.setReplicationSpec(replicationSpec);
+    }
+    desc.getPartition(0).setLocation(ptn.getLocation()); // use existing location
+    return TaskFactory.get(
+        new DDLWork(new HashSet<>(), new HashSet<>(), desc),
+        context.hiveConf
+    );
+  }
+
+  private TaskTracker forExistingTable(String lastPartitionReplicated) throws Exception {
+    boolean encounteredTheLastReplicatedPartition = (lastPartitionReplicated == null);
+    ReplicationSpec replicationSpec = event.replicationSpec();
+    LOG.debug("table partitioned");
+    for (AddPartitionDesc addPartitionDesc : event.partitionDescriptions(tableDesc)) {
```

Comment:
The partitions list from table event metadata is traversed several times in this LoadPartitions task generation flow. Need to optimise this.